[CodeCompletion] Don't suggest argument labels from unapplicable overload

rintaro opened the pull request:
When matching an argument to a parameter list, if there is no matching parameter, the argument list is not applicable to the parameters. In such case, code completion should not suggest argument labels of the function.

rdar://77867723

rintaro committed:
[CodeCompletion] Don't suggest argument labels from unapplicable overload

When matching an argument to a parameter list, if there is no matching
parameter, the argument list is not applicable to the parameters. In
such case, code completion should not suggest argument labels of the
function.

rdar://77867723

rintaro commented:
@swift-ci Please smoke test

ahoppen reviewed (commented):
I don’t think this is an improvement in all cases. I can see that in your test case, it’s definitely what the user is expecting but when completing `self.fn(cc: .up, #^OVERLOAD_LABEL^#)` where the user made a typo and only typed two `c`, we now aren’t providing any results.

I think it would be great if we would fall back to the current behavior, if we don’t get any completions otherwise. An initial idea of mine how to implement it, would be to have a flag on `getPositionInParams` and the main body of `analyzeApplyExpr` (after the comment `// Collect possible types (or labels) at the position.`) toggling the behavior. If that flag is set to “strict argument label matching” and we don’t have any results at the end of `analyzeApplyExpr`, we just invoke the main body again with the flag set to “weak argument label matching”. What do you think?

rintaro committed:
[CodeCompletion] Lenient argument label checking if matching overload not found

If label-matching overload is not found, try to find a position with
lenient label matching.

rintaro commented:
Fair enough. I added a commit to fallback to the existing behavior. WDYT?

rintaro commented:
@swift-ci Please smoke test

ahoppen commented on lib/IDE/ExprContextAnalysis.cpp:
```diff
@@ -772,21 +772,20 @@ static Expr *getArgAtPosition(Expr *Args, unsigned Position) {
 /// in \p Params which don't occur in \p Args.
 ///
 /// \returns \c true if success, \c false if \p CCExpr is not a part of \p Args.
```
The comment now doesn’t match the `Optional` return type anymore…

ahoppen commented on lib/IDE/ExprContextAnalysis.cpp:
```diff
@@ -772,21 +772,20 @@ static Expr *getArgAtPosition(Expr *Args, unsigned Position) {
-                                unsigned &PosInParams) {
+static Optional<unsigned>
+getPositionInParams(DeclContext &DC, Expr *Args, Expr *CCExpr,
+                    ArrayRef<AnyFunctionType::Param> Params, bool lenient) {
```
Nitpick: All the other arguments are capitalized.

ahoppen commented on lib/IDE/ExprContextAnalysis.cpp:
```diff
@@ -947,21 +952,44 @@ class ExprContextAnalyzer {
+                /*lenient=*/true);
+          }
+        }
+      }
```
I think for safety and readability it would be good to have an 
```cpp
assert(posInParams.size() == Candidates.size());
```
here.

ahoppen reviewed (approved):
Looks good to me. Thanks for adding the fallback option.

I’ve got three super minor nitpicks. It would be nice to fix them but if you want to get this merged soon, feel free to ignore them.

rintaro commented:
@swift-ci Please smoke test
